File: Functions/plotCapacityDegradationTile.m
function plotCapacityDegradationTile(axTime, axFEC, combined_df, cell_list, cell_plot_list, colormap_list, Marker_list, plot_title)
% plotCapacityDegradationTile - Draw a relative-capacity-vs-time and
% vs-FEC pair of traces into caller-supplied axes, instead of creating a
% new figure.
%
% Summary: Mirrors the standalone plotCapacityDegradation() figure logic
%          used in PlotCapacityDegradation_detailed.m, but targets
%          externally created axes (axTime, axFEC) so several datasets
%          can be combined as columns/tiles of one shared figure (e.g.
%          via tiledlayout). Legends are intentionally omitted, matching
%          the current no-legend convention already applied to this
%          script's capacity plots.
% Author: NEXTBMS Team
% Date: 2026-08-07
%
% Inputs:
%   axTime         - target axes for the Time [days] vs Relative Capacity subplot
%   axFEC          - target axes for the FEC vs Relative Capacity subplot
%   combined_df    - table with columns cell_number, Timestamp,
%                    "Capacity [Ah]", CheckupCapacityFEC
%   cell_list      - cellstr/string array of all cell numbers, plotted as
%                    thin grey background traces for context
%   cell_plot_list - cellstr/string array of highlighted cell numbers
%   colormap_list  - line color per highlighted cell
%   Marker_list    - line style per highlighted cell
%   plot_title     - base title, suffixed with '(vs Time)' / '(vs FEC)'
%
% Outputs: none (draws directly into axTime/axFEC)

    % --- Time vs Relative Capacity ---
    hold(axTime, 'on');
    % Background: every cell in the dataset, thin grey lines for context.
    for i = 1:length(cell_list)
        cell_name = cell_list{i};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            continue;
        end
        time_days = days(current_cell.("Timestamp") - current_cell.("Timestamp")(1));
        rel_capacity = current_cell.("Capacity [Ah]") / current_cell.("Capacity [Ah]")(1) * 100;
        plot(axTime, time_days, rel_capacity, 'LineWidth', 2, 'Color', [0.5 0.5 0.5 0.2], 'HandleVisibility', 'off');
    end

    % Highlighted cells: colored traces (no legend, per project convention for this script).
    for idx = 1:length(cell_plot_list)
        cell_name = cell_plot_list{idx};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            warning('Cell %s not found in data', cell_name);
            continue;
        end
        time_days = days(current_cell.("Timestamp") - current_cell.("Timestamp")(1));
        rel_capacity = current_cell.("Capacity [Ah]") / current_cell.("Capacity [Ah]")(1) * 100;
        plot(axTime, time_days, rel_capacity, '-o', 'LineWidth', 2, 'Color', colormap_list{idx}, ...
            'MarkerFaceColor', colormap_list{idx}, 'MarkerSize', 4, 'LineStyle', Marker_list{idx}, 'HandleVisibility', 'off');
    end

    grid(axTime, 'on');
    set(axTime, 'GridLineStyle', '--', 'GridAlpha', 0.6, 'FontSize', 11);
    xlabel(axTime, 'Time [days]', 'FontSize', 11);
    ylabel(axTime, 'Relative Capacity [%]', 'FontSize', 11);
    if ~isempty(plot_title)
        title(axTime, [plot_title ' (vs Time)'], 'FontSize', 11);
    end
    xlim(axTime, [0, 425]);
    ylim(axTime, [78, 103]);
    hold(axTime, 'off');

    % --- FEC vs Relative Capacity ---
    hold(axFEC, 'on');
    % Background: every cell in the dataset, thin grey lines for context.
    for i = 1:length(cell_list)
        cell_name = cell_list{i};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            continue;
        end
        rel_capacity = current_cell.("Capacity [Ah]") / current_cell.("Capacity [Ah]")(1) * 100;
        plot(axFEC, current_cell.CheckupCapacityFEC, rel_capacity, 'LineWidth', 2, 'Color', [0.5 0.5 0.5 0.2], 'HandleVisibility', 'off');
    end

    % Highlighted cells: colored traces (no legend, per project convention for this script).
    maxFEC_highlight = 0;
    for idx = 1:length(cell_plot_list)
        cell_name = cell_plot_list{idx};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            continue;
        end
        rel_capacity = current_cell.("Capacity [Ah]") / current_cell.("Capacity [Ah]")(1) * 100;
        plot(axFEC, current_cell.CheckupCapacityFEC, rel_capacity, '-o', 'LineWidth', 2, 'Color', colormap_list{idx}, ...
            'MarkerFaceColor', colormap_list{idx}, 'MarkerSize', 4, 'LineStyle', Marker_list{idx}, 'HandleVisibility', 'off');
        if ~isempty(current_cell.CheckupCapacityFEC)
            maxFEC_highlight = max([maxFEC_highlight; max(current_cell.CheckupCapacityFEC)]);
        end
    end

    grid(axFEC, 'on');
    set(axFEC, 'GridLineStyle', '--', 'GridAlpha', 0.6, 'FontSize', 11);
    xlabel(axFEC, 'Full Equivalent Cycles (FEC)', 'FontSize', 11);
    ylabel(axFEC, 'Relative Capacity [%]', 'FontSize', 11);
    if ~isempty(plot_title)
        title(axFEC, [plot_title ' (vs FEC)'], 'FontSize', 11);
    end
    ylim(axFEC, [78, 103]);
    xlim(axFEC, [0, maxFEC_highlight]);
    hold(axFEC, 'off');
end

File: Functions/plotCapacityDegradationTile.py
"""
plotCapacityDegradationTile - Draw a relative-capacity-vs-time and
relative-capacity-vs-FEC pair of traces into caller-supplied matplotlib axes.

Python counterpart of Functions/plotCapacityDegradationTile.m, used by
python_scripts/PlotCapacityDegradation_detailed.py (todo #018).

Mirrors the MATLAB tile helper exactly: draws every cell in cell_list as a
faint grey background trace (2 pt line, alpha 0.2), then the highlighted
cells in cell_plot_list as colored 'o'-marker traces. No legend is drawn on
either axes - this matches the current no-legend convention already applied
to every capacity-degradation figure in the MATLAB script.

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-25

Compliance: R-004, R-013, R-016.
"""


def plot_capacity_degradation_tile(ax_time, ax_fec, combined_df, cell_list,
                                    cell_plot_list, colormap_list, marker_list, plot_title):
    """
    Draw the Time [days] / FEC relative-capacity trace pair into ax_time/ax_fec.

    Parameters
    ----------
    ax_time, ax_fec : matplotlib.axes.Axes
        Target axes for the Time [days] and FEC subplots respectively.
    combined_df : pandas.DataFrame
        Table with columns cell_number, Timestamp, "Capacity [Ah]",
        CheckupCapacityFEC.
    cell_list : list of str
        All cell numbers in the dataset, drawn as thin grey background
        traces for context.
    cell_plot_list : list of str
        Highlighted cell numbers to draw in color.
    colormap_list : list
        Line color per highlighted cell (matplotlib color spec).
    marker_list : list of str
        Line style per highlighted cell (e.g. '-', '--').
    plot_title : str
        Base title, suffixed with '(vs Time)' / '(vs FEC)'. Pass '' for no title.

    Returns
    -------
    None (draws directly into ax_time/ax_fec; no legend is created).
    """

    # --- Time vs Relative Capacity ---
    # Background: every cell in the dataset, thin grey lines for context (no legend).
    for cell_name in cell_list:
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            continue
        current_cell = current_cell.sort_values("Timestamp")
        time_days = (current_cell["Timestamp"] - current_cell["Timestamp"].iloc[0]).dt.total_seconds() / 86400.0
        rel_capacity = current_cell["Capacity [Ah]"] / current_cell["Capacity [Ah]"].iloc[0] * 100
        ax_time.plot(time_days, rel_capacity, linewidth=2, color=(0.5, 0.5, 0.5, 0.2))

    # Highlighted cells: colored 'o'-marker traces (still no legend).
    for idx, cell_name in enumerate(cell_plot_list):
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            print(f"Warning: Cell {cell_name} not found in data")
            continue
        current_cell = current_cell.sort_values("Timestamp")
        time_days = (current_cell["Timestamp"] - current_cell["Timestamp"].iloc[0]).dt.total_seconds() / 86400.0
        rel_capacity = current_cell["Capacity [Ah]"] / current_cell["Capacity [Ah]"].iloc[0] * 100
        ax_time.plot(time_days, rel_capacity, marker="o", linewidth=2, color=colormap_list[idx],
                     markerfacecolor=colormap_list[idx], markersize=4, linestyle=marker_list[idx])

    ax_time.grid(True, linestyle="--", alpha=0.6)
    ax_time.tick_params(labelsize=11)
    ax_time.set_xlabel("Time [days]", fontsize=11)
    ax_time.set_ylabel("Relative Capacity [%]", fontsize=11)
    if plot_title:
        ax_time.set_title(f"{plot_title} (vs Time)", fontsize=11)
    ax_time.set_xlim(0, 425)
    ax_time.set_ylim(78, 103)

    # --- FEC vs Relative Capacity ---
    for cell_name in cell_list:
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            continue
        rel_capacity = current_cell["Capacity [Ah]"] / current_cell["Capacity [Ah]"].iloc[0] * 100
        ax_fec.plot(current_cell["CheckupCapacityFEC"], rel_capacity, linewidth=2, color=(0.5, 0.5, 0.5, 0.2))

    max_fec_highlight = 0.0
    for idx, cell_name in enumerate(cell_plot_list):
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            continue
        rel_capacity = current_cell["Capacity [Ah]"] / current_cell["Capacity [Ah]"].iloc[0] * 100
        ax_fec.plot(current_cell["CheckupCapacityFEC"], rel_capacity, marker="o", linewidth=2,
                    color=colormap_list[idx], markerfacecolor=colormap_list[idx], markersize=4,
                    linestyle=marker_list[idx])
        if current_cell["CheckupCapacityFEC"].notna().any():
            max_fec_highlight = max(max_fec_highlight, float(current_cell["CheckupCapacityFEC"].max()))

    ax_fec.grid(True, linestyle="--", alpha=0.6)
    ax_fec.tick_params(labelsize=11)
    ax_fec.set_xlabel("Full Equivalent Cycles (FEC)", fontsize=11)
    ax_fec.set_ylabel("Relative Capacity [%]", fontsize=11)
    if plot_title:
        ax_fec.set_title(f"{plot_title} (vs FEC)", fontsize=11)
    ax_fec.set_ylim(78, 103)
    ax_fec.set_xlim(0, max_fec_highlight)


__all__ = ["plot_capacity_degradation_tile"]
